### ISSUE TITLE
Revise jack driver

### ISSUE THREAD

**Issue by derselbst** [FluidSynth/fluidsynth#416]
The implementation of fluid_jack.c has many open TODOs and needs to be revised. Potentially also taking #79  into account.

**Comment by derselbst** [comment]
It seems that the current implementation is not able to start two audio jack drivers: A first jack client is created, but the second client simply reuses the first client. The output_ports of the first client have already been setup, but they will be set up again by the second call, which is not possible as they already exist, thus `jack_port_register()` returns NULL, essentially overwriting the correctly set up ports. This will then cause the `jack_port_get_buffer()` to return a NULL buffer, which will be passed to `fluid_synth_write_float()`. Crash.

**Comment by rncbc** [comment]
I can confirm that the current implementation (as of released V2.0.0) is not able to start or run more than one audio jack driver; for instance Qsynth crashes/segfaults when trying to create a second engine.

This issue is currently a qsynth-git (on libfluidsynth2) showstopper, at least for the folks that have premore than one qsynth engine working fine for ages (on libfluidsynth1).

**Comment by derselbst** [comment]
But I cant think of any significant change to the jack driver that explains why it worked in fluidsynth1. And it seems not that trivial to fix either :/

**Comment by mawe42** [comment]
I could recreate the crashes again, but only when I choose a fairly large SF3 soundfont for one of the synths, like the MuseScore_General.sf3. Smaller SF3 files don't seem to trigger the crash. Could it be a timing/locking issue?

**Comment by rncbc** [comment]
seems that having Options > Output peak level meters turned on makes all the difference...
```
[Options]
OutputMeters=false
```
when set to true it crashes on sight of a second engine.

**Comment by mawe42** [comment]
I'm not sure it's even Qsynth related. I can reproduce the crash with fluidsynth, a Jack server running and loading an SF3 file on the command line. It crashes here:
https://github.com/FluidSynth/fluidsynth/blob/f4b88fc7471e16d73be8f2f1a495ef180fdf304c/src/drivers/fluid_jack.c#L207

When specifying a new server name with audio.jack.server it uses the jack_client_open above that and everything works fine:
https://github.com/FluidSynth/fluidsynth/blob/f4b88fc7471e16d73be8f2f1a495ef180fdf304c/src/drivers/fluid_jack.c#L203

**Comment by mawe42** [comment]
I've got the feeling it's connected to the new sample loading mechanism. It crashes with SF2 and dynamic-sample-loading enabled, and with SF3 regardless of dynamic-sample-loading. Just not with a very small SF3 file. This is really strange...

**Comment by rncbc** [comment]
i can confirm the finding in https://github.com/FluidSynth/fluidsynth/issues/416#issuecomment-406779147, a second JACK client is not created at least with the given client name and no new JACK client/ports are registered, whenever a second Qsynth engine is setup to "jack" audio driver, even thought output peak lvel meters are turned off--if turned on, it crashes, most probably on the second bogus client's `fluid_jack_driver_process() `... just sayin' :)

**Comment by rncbc** [comment]
> As @derselbst has noticed, there's a bug in jack2 where failed mlock calls are not properly handled,

FWIW: i've tested on jack2 (1.9.12) and jack1 (0.125.0) with either "No memory lock" option on and off  (-m): no difference.

**Comment by mawe42** [comment]
> FWIW: i've tested on jack2 (1.9.12) and jack1 (0.125.0) with either "No memory lock" option on and off (-m): no difference.

The "No memory lock" option doesn't seem to disable the memory lock in the place that is triggering the bug. It's a memlock attempted in an mmap call, for a mutex:
https://github.com/jackaudio/jack2/commit/cc8576a7ab62caab10c7ddd7309367cf797c5e69

**Comment by mawe42** [comment]
And jack1 had at least one very similar bug, fixed in
https://github.com/jackaudio/jack1/commit/0326eb4afeac07df7d99bed844c4243bbf9c78a2

**Comment by derselbst** [comment]
After I've patched jack to correctly check for mmap(), I started qsynth again, attempting to run >10 instances. 

> when starting with qsynth -o audio.jack.server=default, a second JACK client engine gets created and ports registered alright

Doesnt work for me. Only the first qsynth instance is able to create a driver. Subsequent instances fail to create the audio driver: `fluidsynth: error: Failed to connect to Jack server.`

Weird issue indeed. Necessitates some serious investigation, which I'm currently lacking time for.
